tests/azure_iot_provisioning_client_ut.c

```diff
@@ -0,0 +1,704 @@
+                                                       &ucEndpoint[ 0 ], sizeof( ucEndpoint ),
+                                                       &ucIdScope[ 0 ], sizeof( ucIdScope ),
+                                                       &ucRegistrationId[ 0 ], sizeof( ucRegistrationId ),
+                                                       NULL, ucBuffer, sizeof( ucBuffer ),
+                                                       prvGetUnixTime,
+                                                       &xTransportInterface ),
+                      eAzureIoTProvisioningSuccess );
+
+    AzureIoTProvisioningClient_Deinit( &xTestProvisioningClient );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTProvisioningClient_SymmetricKeySet_Failure( void ** ppvState )
+{
+    AzureIoTProvisioningClient_t xTestProvisioningClient;
+
+    ( void ) ppvState;
+
+    /* Fail AzureIoTProvisioningClient_SetSymmetricKey when null symmetric key is passed */
+    assert_int_not_equal( AzureIoTProvisioningClient_SetSymmetricKey( &xTestProvisioningClient,
+                                                                      NULL, 0,
+                                                                      prvHmacFunction ),
+                          eAzureIoTProvisioningSuccess );
+
+    /* Fail AzureIoTProvisioningClient_SetSymmetricKey when null hashing function is passed */
+    assert_int_not_equal( AzureIoTProvisioningClient_SetSymmetricKey( &xTestProvisioningClient,
+                                                                      ucSymmetricKey, sizeof( ucSymmetricKey ) - 1,
+                                                                      NULL ),
+                          eAzureIoTProvisioningSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTProvisioningClient_SymmetricKeySet_Success( void ** ppvState )
+{
+    AzureIoTProvisioningClient_t xTestProvisioningClient;
+
+    ( void ) ppvState;
+
+    assert_int_equal( AzureIoTProvisioningClient_SetSymmetricKey( &xTestProvisioningClient,
+                                                                  ucSymmetricKey, sizeof( ucSymmetricKey ) - 1,
+                                                                  prvHmacFunction ),
+                      eAzureIoTProvisioningSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTProvisioningClient_Register_ConnectFailure( void ** ppvState )
+{
+    AzureIoTProvisioningClient_t xTestProvisioningClient;
+    AzureIoTProvisioningClientResult_t xResult;
+
+    ( void ) ppvState;
+
+    prvSetupTestProvisioningClient( &xTestProvisioningClient );
+
+    /* Connect */
+    will_return( prvHmacFunction, 0 );
+    will_return( AzureIoTMQTT_Connect, eAzureIoTMQTTServerRefused );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
```

Comment:
I didn't pick up on this in product code review, but if the _Connect call fails should we even try to do a ProcessLoop?  It seems we should bail out since I can't see how Process would succeed. ?

Reply:
Yes, right now current impl trigger Action and then we always  process loop irrespective of action failed or not. 

I can update the impl.

---
In reply to: [607965358](https://github.com/Azure/azure-iot-middleware-freertos/pull/62#discussion_r607965358) [](ancestors = 607965358)